Fix bug which is it doesn't raise exception if resource is not found

camelmasa opened the pull request:
Fixed https://github.com/CanCanCommunity/cancancan/issues/456

camelmasa commented on spec/cancan/controller_resource_spec.rb:
```diff
@@ -501,10 +501,10 @@ class Model < ::Model; end
-      resource = CanCan::ControllerResource.new(controller, find_by: :find_by_name)
+      resource = CanCan::ControllerResource.new(controller, find_by: :name)
       resource.load_resource
       expect(controller.instance_variable_get(:@model)).to eq(model)
```
This test case was not worked before. It was same as https://github.com/CanCanCommunity/cancancan/pull/461/files#diff-eba4e6277d1f300a2baa8a35cef7ef0fR492.

so I updated test case.

coorasse replied:
this is actually a bit misleading. could you restore the intention of the previous test? but maybe use `allow(Model).to receive(:search).with('foo') { model }` on line 504 and on line 507 you use `find_by: :search` which should reflect the previous behaviour.

camelmasa replied:
Ah I see. I didn't notice that. Got it.

camelmasa commented on lib/cancan/controller_resource_finder.rb:
```diff
@@ -13,7 +13,9 @@ def find_resource
-      if resource_base.respond_to? 'find_by'
+      if resource_base.respond_to? "find_by_#{@options[:find_by]}!"
+        resource_base.send("find_by_#{@options[:find_by]}!", id_param)
+      elsif resource_base.respond_to? 'find_by'
```
Revived it from here https://github.com/CanCanCommunity/cancancan/commit/33df7eb51a85692e702dc0e2a7252471bfe271ef#diff-38607cbe9e3608bd14d9367c9cf91eefL110

lostapathy commented on lib/cancan/controller_resource_finder.rb:
```diff
@@ -13,7 +13,9 @@ def find_resource
+      if resource_base.respond_to? "find_by_#{@options[:find_by]}!"
+        resource_base.send("find_by_#{@options[:find_by]}!", id_param)
+      elsif resource_base.respond_to? 'find_by'
         resource_base.send('find_by', @options[:find_by].to_sym => id_param)
```
I know it didn't start in this PR, but why are we calling .send() on this rather than going direct in this case, such as:

resource_base.find_by(@options[:find_by].to_sym => id_param)

camelmasa replied:
Right, I think we can call `find_by` directly like that 👍  That would be great.

coorasse commented on lib/cancan/controller_resource_finder.rb:
```diff
@@ -13,8 +13,10 @@ def find_resource
     def find_resource_using_find_by
-      if resource_base.respond_to? 'find_by'
-        resource_base.send('find_by', @options[:find_by].to_sym => id_param)
+      if resource_base.respond_to? "find_by_#{@options[:find_by]}!"
```
support for dynamic finders has been dropped.

coorasse commented on lib/cancan/controller_resource_finder.rb:
```diff
@@ -13,8 +13,10 @@ def find_resource
+      if resource_base.respond_to? "find_by_#{@options[:find_by]}!"
+        resource_base.send("find_by_#{@options[:find_by]}!", id_param)
+      elsif resource_base.respond_to? 'find_by'
+        resource_base.find_by(@options[:find_by].to_sym => id_param)
```
👍

coorasse commented on spec/cancan/controller_resource_spec.rb.orig:
```diff
@@ -1,708 +0,0 @@
-require 'spec_helper'
```
how did this end up there? 😅

camelmasa replied:
I found that on  https://github.com/CanCanCommunity/cancancan/pull/452/commits/1e6b5705d1a57b5e49faa1a2732291802ba72be9 😃

coorasse commented:
Please look at my comment: https://github.com/CanCanCommunity/cancancan/issues/456#issuecomment-346081557 I think I misinterpreted partially your PR while reviewing it

camelmasa commented:
@coorasse  Thanks for comments. I updated this PR. Could you review when you have time ❓ 

Also, I changed one test case on https://github.com/CanCanCommunity/cancancan/pull/461/commits/5c9d21df8c18e1befaa3699e72032cc04eafa841.
Is my understanding correct ❓

coorasse commented:
Following @terryyin comment (https://github.com/CanCanCommunity/cancancan/commit/8fc2a2dbefc542749650da6b957335e8075b34a3#commitcomment-25775734) we have a regression at the moment. I am restoring the old behaviour now through another PR and then we can look at a refactoring which would deprecate the existing behaviour and therefore fall into a major release

coorasse commented:
could you please check your PR and update it against the current develop branch? Thanks

terryyin commented:
The dynamic finders with ! is still supported in Rails 5 (http://edgeguides.rubyonrails.org/active_record_querying.html#dynamic-finders). Perhaps there's no need to change this to use "find_by!" now, because it still changes the current behaviour (people might have their dynamic finder override).

camelmasa commented:
@terryyin Thanks for advice 😃 

@coorasse I confirmed latest version 2.1.2 is released. I'll close it. Thanks for releasing.

terryyin commented:
Thanks!

Great work, BTW.

> On 23 Nov 2017, at 11:02 AM, Masahiro Saito <notifications@github.com> wrote:
> 
> @terryyin <https://github.com/terryyin> Thanks for advice 😃
> 
> @coorasse <https://github.com/coorasse> I confirmed latest version 2.1.2 is released. I'll close it. Thanks for releasing.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/CanCanCommunity/cancancan/pull/461#issuecomment-346522967>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAwJYsh_SqqRHmBnu4SNEJtleYa4jr9dks5s5OBSgaJpZM4Qk3q4>.
>

lostapathy commented:
Thanks @coorasse for taking care of this
